File: app/src/main/java/mk/webfactory/template/network/OAuthInterceptor.java
package mk.webfactory.template.network;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Function;
import java.io.IOException;
import mk.webfactory.template.data.rx.Observables;
import mk.webfactory.template.data.rx.StubDisposableObserver;
import mk.webfactory.template.model.auth.AccessToken;
import mk.webfactory.template.user.UserDataWrapper;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;

final class OAuthInterceptor implements Interceptor {

  private AccessToken accessToken;
  private Disposable userUpdateStreamDisposable;

  OAuthInterceptor() {}

  OAuthInterceptor(AccessToken accessToken) {
    this.accessToken = accessToken;
  }

  void setUserUpdateStream(@NonNull Observable<UserDataWrapper> userUpdateStream) {
    Observables.dispose(userUpdateStreamDisposable);
    userUpdateStreamDisposable = userUpdateStream
        .map(new Function<UserDataWrapper, AccessToken>() {
          @Override public AccessToken apply(UserDataWrapper userDataWrapper) throws Exception {
            return userDataWrapper.getAccessToken();
          }
        })
        .subscribeWith(new StubDisposableObserver<AccessToken>() {
          @Override public void onNext(final AccessToken accessToken) {
            if (AccessToken.VOID.equals(accessToken)) {
              updateAccessToken(null);
            } else {
              updateAccessToken(accessToken);
            }
          }
        });
  }

  void disposeUserUpdateStream() {
    Observables.dispose(userUpdateStreamDisposable);
  }

  private synchronized void updateAccessToken(@Nullable AccessToken accessToken) {
    this.accessToken = accessToken;
  }

  private synchronized AccessToken getAccessToken() {
    return accessToken;
  }

  @Override public Response intercept(@NonNull Chain chain) throws IOException {
    final AccessToken accessToken = getAccessToken();
    if (accessToken != null) {
      Request request = chain.request().newBuilder()
          .header("Authorization", "Bearer " + accessToken.getToken())
          .build();
      return chain.proceed(request);
    } else {
      return chain.proceed(chain.request());
    }
  }
}

File: app/src/main/java/mk/webfactory/template/network/OAuthInterceptor.kt
package mk.webfactory.template.network

import io.reactivex.Observable
import io.reactivex.disposables.Disposable
import io.reactivex.internal.disposables.DisposableHelper.dispose
import mk.webfactory.template.model.auth.AccessToken
import mk.webfactory.template.model.user.UserSession
import okhttp3.Interceptor
import okhttp3.Response
import java.io.IOException
import java.util.concurrent.atomic.AtomicReference

/**
 *
 * [Interceptor] that adds OAuth2 authorization header on each request.
 *
 * Mind to update the token value when it changes.
 * Setting it to `null` disables this interceptor.
 */

class OAuthInterceptor() : Interceptor {
    @get:Synchronized private var accessToken: AccessToken? = null
    private lateinit var userUpdateStreamDisposable: Disposable

    fun setUserUpdateStream(userUpdateStream: Observable<UserSession?>) {
        dispose(AtomicReference(userUpdateStreamDisposable))
        userUpdateStreamDisposable = userUpdateStream
            .map { userSession -> userSession.accessToken }
            .subscribe { t ->
                if (t != accessToken) {
                    updateAccessToken(t)
                }
            }
    }

    @Synchronized
    private fun updateAccessToken(accessToken: AccessToken?) {
        this.accessToken = accessToken
    }

    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {
        val accessToken = accessToken
        return if (accessToken != null) {
            val request = chain.request().newBuilder()
                .header("Authorization", "Bearer " + accessToken.token)
                .build()
            chain.proceed(request)
        } else {
            chain.proceed(chain.request())
        }
    }
}